File: src/main/scala/ex3/List.scala
package ex3

import scala.annotation.tailrec

enum List[+A]:
  case Nil extends List[Nothing]
  case Cons(head: A, tail: List[A])

object List:
  def sum(ints: List[Int]): Int = ints match
    case Nil => 0
    case Cons(x, xs) => x + sum(xs)

  def product(ds: List[Double]): Double = ds match
    case Nil => 1.0
    case Cons(0.0, _) => 0.0
    case Cons(x, xs) => x * product(xs)

  def apply[A](as: A*): List[A] =
    if as.isEmpty then Nil
    else Cons(as.head, apply(as.tail*))

  def foldRight[A, B](l: List[A], z: B)(f: (A, B) => B): B = l match
    case Nil => z
    case Cons(x, xs) => f(x, foldRight(xs, z)(f))

  def append[A](l: List[A], r: List[A]): List[A] = l match
    case Nil => r
    case Cons(h, t) => Cons(h, append(t, r))

  // ex 3.2
  def tail[A](l: List[A]): List[A] = l match
    case Nil => Nil
    case Cons(_, t) => t

  // ex 3.3
  def setHead[A](l: List[A], h: A): List[A] = l match
    case Nil => Cons(h, Nil)
    case Cons(_, t) => Cons(h, t)

  // ex 3.4
  @tailrec
  def drop[A](l: List[A], n: Int): List[A] =
    if n == 0 then l
    else l match
      case Nil => Nil
      case Cons(_, t) => drop(t, n - 1)

  // ex 3.5
  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = l match
    case Cons(h, t) => if f(h) then dropWhile(t, f) else l
    case Nil => Nil

  // ex 3.6
  def init[A](l: List[A]): List[A] = l match
    case Nil => Nil
    case Cons(_, Nil) => Nil
    case Cons(h, t) => Cons(h, init(t))

  // ex 3.9
  def length[A](l: List[A]): Int = l match
    case Nil => 0
    case Cons(_, t) => 1 + length(t)

  // ex 3.10
  def foldLeft[A,B](l: List[A], z: B)(f: (B, A) => B): B = l match
    case Nil => z
    case Cons(h, t) => foldLeft(t, f(z, h))(f)


  // ex 3.11
  def sumViaFoldLeft(l: List[Int]): Int =
    foldLeft(l, 0)(_ + _)

  def productViaFoldLeft(l: List[Double]): Double =
    foldLeft(l, 1.0)(_ * _)

  // ex 3.12
  def reverse[A](l: List[A]): List[A] =
    foldLeft(l, Nil: List[A])((acc, emt) => Cons(emt, acc))

  // ex 3.13
  def foldLeftViaFoldRight[A,B](l: List[A], z: B)(f: (B, A) => B): B =
    foldRight(l, (b: B) => b)((a, idf) => b => idf(f(b, a)))(z)

  def foldRightViaFoldLeft[A,B](l: List[A], z: B)(f: (A, B) => B): B =
    foldLeft(l, (b: B) => b)((idf, a) => b => idf(f(a, b)))(z)

  // ex 3.14
  def appendViaFoldRight[A](l: List[A], r: List[A]): List[A] =
    foldRight(l, r)(Cons(_, _))

  // ex 3.15
  def concat[A](l: List[List[A]]): List[A] =
    foldRight(l, Nil: List[A])(append)

  // ex 3.16
  def addOneAllElements(l: List[Int]): List[Int] =
    foldRight(l, Nil: List[Int])((h, t) => Cons(h + 1, t))

  // ex 3.17
  def doubleToString(l: List[Double]): List[String] =
    foldRight(l, Nil: List[String])((h, t) => Cons(h.toString, t))

  // ex 3.18
  def map[A,B](l: List[A])(f: A => B): List[B] =
    foldRight(l, Nil: List[B])((h, t) => Cons(f(h), t))

  // ex 3.19
  def filter[A](l: List[A])(f: A => Boolean): List[A] =
    foldRight(l, Nil: List[A])((h, t) => if f(h) then Cons(h, t) else t)

  // ex 3.20
  def flatMap[A,B](l: List[A])(f: A => List[B]): List[B] =
    concat(map(l)(f))

  // ex 3.21
  def filterViaFlatMap[A](l: List[A])(f: A => Boolean): List[A] =
    flatMap(l)(a => if f(a) then List(a) else Nil)

  // ex 3.22
  def addPairwise(a: List[Int], b: List[Int]): List[Int] = (a, b) match
    case (Nil, _) => Nil
    case (_, Nil) => Nil
    case (Cons(h1, t1), Cons(h2, t2)) => Cons(h1 + h2, addPairwise(t1, t2))

  // ex 3.23
  def zipWith[A,B,C](a: List[A], b: List[B])(f: (A,B) => C): List[C] = (a, b) match
    case (Nil, _) => Nil
    case (_, Nil) => Nil
    case (Cons(h1, t1), Cons(h2, t2)) => Cons(f(h1, h2), zipWith(t1, t2)(f))

  // ex 3.24
  @tailrec
  def hasSubsequence[A](sup: List[A], sub: List[A]): Boolean = (sup, sub) match
    case (_, Nil) => true
    case (Nil, _) => false
    case (Cons(h1, t1), Cons(h2, t2)) =>
      if h1 == h2 then hasSubsequence(t1, t2)
      else hasSubsequence(t1, sub)

File: FunctionalPrograming/src/main/kotlin/ex3/List.kt
package org.example.ex3

sealed class List<out A> {
    companion object {
        fun <A> of(vararg aa: A): List<A> {
            val tail = aa.sliceArray(1 until aa.size)
            return if (aa.isEmpty()) Nil else Cons(aa[0], of(*tail))
        }

        fun sum(ints: List<Int>): Int =
            foldRight(ints, 0) { a, b -> a + b }

        fun product(dbs: List<Double>): Double =
            foldRight(dbs, 1.0) { a, b -> a * b }


        fun <A, B> foldRight(xs: List<A>, z: B, f: (A, B) -> B): B = when (xs) {
            is Nil -> z
            is Cons -> f(xs.head, foldRight(xs.tail, z, f))
        }


        fun <A> append(a1: List<A>, a2: List<A>): List<A> = when (a1) {
            is Nil -> a2
            is Cons -> Cons(a1.head, append(a1.tail, a2))
        }

        fun <A> empty(): List<A> = Nil

        // 3.1
        fun <A> tail(xs: List<A>): List<A> = when (xs) {
            is Nil -> Nil
            is Cons -> xs.tail
        }

        // 3.2
        fun <A> setHead(xs: List<A>, x: A): List<A> = when (xs) {
            is Nil -> Cons(x, Nil)
            is Cons -> Cons(x, xs)
        }

        // 3.3
        fun <A> drop(l: List<A>, n: Int): List<A> = when (l) {
            is Nil -> Nil
            is Cons -> if (n == 0) l else drop(l.tail, n - 1)
        }

        // 3.4
        fun <A> dropWhile(l: List<A>, f: (A) -> Boolean): List<A> = when (l) {
            is Nil -> Nil
            is Cons -> if (f(l.head)) dropWhile(l.tail, f) else l
        }

        // 3.5
        fun <A> init(l: List<A>): List<A> = when (l) {
            is Nil -> Nil
            is Cons -> if (l.tail == Nil) Nil else Cons(l.head, init(l.tail))
        }

        // 3.8
        fun <A> length(xs: List<A>): Int =
            foldRight(xs, 0) { _, acc -> acc + 1 }

        // 3.9
        tailrec fun <A, B> foldLeft(xs: List<A>, z: B, f: (B, A) -> B): B = when (xs) {
            is Nil -> z
            is Cons -> foldLeft(xs.tail, f(z, xs.head), f)
        }

        // 3.10
        fun sumViaFoldLeft(xs: List<Int>): Int =
            foldLeft(xs, 0) { a, b -> a + b }

        fun productViaFoldLeft(xs: List<Double>): Double =
            foldLeft(xs, 1.0) { a, b -> a * b }

        fun <A> lengthViaFoldLeft(xs: List<A>): Int =
            foldLeft(xs, 0) { acc, _ -> acc + 1 }

        // 3.11
        fun reverse(xs: List<Int>): List<Int> =
            foldLeft(xs, empty()) { acc, x -> Cons(x, acc) }

        // 3.12
        fun <A, B> foldRightViaFoldLeft(xs: List<A>, z: B, f: (A, B) -> B): B =
            foldLeft(xs, { x:B -> x } ) { idf, a -> { b -> idf(f(a, b)) }  } (z)

        // 3.13
        fun <A> appendViaFoldRight(a1: List<A>, a2: List<A>): List<A> =
            foldRightViaFoldLeft(a1, a2) { x, acc -> Cons(x, acc) }

        // 3.14
        fun <A> concat(xs: List<List<A>>): List<A> =
            foldRightViaFoldLeft(xs, empty(), ::appendViaFoldRight)

        // 3.15
        fun addOneAllElements(xs: List<Int>): List<Int> =
            foldRightViaFoldLeft(xs, empty()) { x, acc -> Cons(x + 1, acc) }

        // 3.16
        fun doubleToString(xs: List<Double>): List<String> =
            foldRightViaFoldLeft(xs, empty()) { x, acc -> Cons(x.toString(), acc) }

        // 3.17
        fun <A, B> map(xs: List<A>, f: (A) -> B): List<B> =
            foldRightViaFoldLeft(xs, empty()) { x, acc -> Cons(f(x), acc) }

        // 3.18
        fun <A> filter(xs: List<A>, f: (A) -> Boolean): List<A> =
            foldRightViaFoldLeft(xs, empty()) { x, acc -> if (f(x)) Cons(x, acc) else acc }

        // 3.19
        fun <A, B> flatMap(xs: List<A>, f: (A) -> List<B>): List<B> =
            concat(map(xs, f))

        // 3.20
        fun <A> filterViaFlatMap(xs: List<A>, f: (A) -> Boolean): List<A> =
            flatMap(xs) { if (f(it)) List.of(it) else empty() }
    }
}

data object Nil: List<Nothing>()

data class Cons<A>(val head: A, val tail: List<A>): List<A>()
fun <A> List<A>.tail(): List<A> = List.tail(this)